extmod/btstack: Schedule ops for background completion.

jimmo opened the pull request:
The goal of this PR is to allow using `ble.gatts_notify()` at any time, even if the stack is not ready to send the notification right now.

This also addresses the same issue for `ble.gatts_indicate` and `ble.gattc_write` (without response). The result is that you can now:
 - Call notify/indicate at any time.
 - Call write-without-response while another one is in progress. (But no additional ones can queue, and will fail with EALREADY)

Unfortunately the three cases notify/indicate, write-with-response and write-without-response all require slightly different implementations.

This PR also addresses two other issues:
 - The buffer passed to write-with-response wasn't copied, meaning it could be modified by the caller, affecting the in-progress write.
 - The buffer wasn't tracked to prevent the GC free'ing it.

jimmo committed:
extmod/btstack: Schedule ops for background completion.

gatts_notify/indicate will now run in the background if the ACL buffer is
currently full.

gattc_write(mode=0) (no response) will now allow for one outstanding write.

gattc_write(mode=1) (with response) will now copy the buffer so that it can't
be modified by the caller.

All four paths also now track the buffer while the op is in progress.

jimmo committed:
examples/bluetooth/ble_advertising.py: Skip appearance field if not set.

jimmo committed:
examples/bluetooth: Add simple UART demo.

jimmo committed:
ports/unix: Make the ATOMIC_SECTION mutex recursive.

This mutex is used to make the unix port behave more like bare metal (i.e. it
allows "IRQ handlers" to run exclusively).

In order to match the bare-metal behavior of MICROPY_BEGIN_ATOMIC_SECTION,
this mutex needs to be recursive.

dpgeorge commented on ports/unix/mpthreadport.c:
```diff
@@ -65,7 +65,8 @@ STATIC pthread_key_t tls_key;
 // But also, e.g. scheduler state.
-STATIC pthread_mutex_t thread_mutex = PTHREAD_MUTEX_INITIALIZER;
+STATIC pthread_mutex_t thread_mutex;
+STATIC pthread_mutexattr_t thread_mutex_attr;
```
Does this attr struct need to be in global data?  Can't it be temporary, on the stack in `mp_thread_init()`?  Is there any reference that says either way?

jimmo replied:
Looking at the implementation, yes, seems fine to put it on the stack. Also, there's an example of doing exactly that in my copy of APUE.

dpgeorge commented on extmod/modbluetooth.c:
```diff
@@ -1147,47 +1147,46 @@ mp_bluetooth_gatts_db_entry_t *mp_bluetooth_gatts_db_lookup(mp_gatts_db_t db, ui
-    }
+    if (entry) {
+        if (value_len > entry->data_alloc) {
+            entry->data = m_new(uint8_t, value_len);
```
This may raise an exception which would leave the function without releasing the bluetooth lock (yes this would be much clearer without NLR...).  I think it can be fixed pretty easily by using `m_new_maybe()` and then explicitly raising an OOM error after releasing the lock.  Or just returning `MP_ENOMEM`.  But the latter requires all callers to check the return value if they can't guarantee the write won't need to allocate.

Also, related, do we need to check that all calls of `mp_bluetooth_gatts_db_write()` do not have the bluetooth lock?

dpgeorge replied:
> Also, related, do we need to check that all calls of `mp_bluetooth_gatts_db_write()` do not have the bluetooth lock?

I checked this.  It looks ok, the bluetooth lock is only held in sections of code that are "low level" or call into BTstack.  Ie there are no calls into the uPy runtime (that could raise).

jimmo replied:
There are very few callers of these methods, and most already checked the return value. I've made all uses do so.

dpgeorge commented on extmod/modbluetooth.c:
```diff
@@ -1147,47 +1147,46 @@ mp_bluetooth_gatts_db_entry_t *mp_bluetooth_gatts_db_lookup(mp_gatts_db_t db, ui
-    if (!entry) {
-        return MP_EINVAL;
+    if (entry) {
+        entry->data = m_renew(uint8_t, entry->data, entry->data_alloc, len);
```
This will have a similar problem, and solution, as `mp_bluetooth_gatts_db_write()`.

dpgeorge commented on extmod/btstack/modbluetooth_btstack.h:
```diff
@@ -33,6 +33,11 @@
+typedef struct _mp_btstack_pending_op_t mp_btstack_pending_op_t;
+
+//typedef struct btstack_linked_item btstack_linked_item_t;
+//typedef btstack_linked_item_t * btstack_linked_list_t;
```
These 3 can be removed.

dpgeorge commented on extmod/btstack/modbluetooth_btstack.c:
```diff
@@ -85,6 +83,148 @@ STATIC mp_obj_bluetooth_uuid_t create_mp_uuid(uint16_t uuid16, const uint8_t *uu
+//  - Holds a GC reference to the copied outgoing buffer.
+//  - Provides enough information for the callback handler to execute the desired operation.
+struct _mp_btstack_pending_op_t {
+    mp_btstack_pending_op_t *next; // Must be first field to match btstack_linked_item.
```
Following BTstack, and to make it clear this struct is a valid linked-list, this should probably be `btstack_linked_item_t *next;`

dpgeorge commented on extmod/btstack/modbluetooth_btstack.c:
```diff
@@ -85,6 +83,148 @@ STATIC mp_obj_bluetooth_uuid_t create_mp_uuid(uint16_t uuid16, const uint8_t *uu
     return result;
 }
 
+// Notes on supporting background ops (i.e. an attempt to gatts_notify while and existing
```
ie -> eg?  and->an

dpgeorge commented on extmod/btstack/modbluetooth_btstack.c:
```diff
@@ -85,6 +83,148 @@ STATIC mp_obj_bluetooth_uuid_t create_mp_uuid(uint16_t uuid16, const uint8_t *uu
+// * Otherwise, fails with GATT_CLIENT_BUSY.
+// * Use gatt_client_request_can_write_without_response_event to get callback
+//   * Takes btstack_packet_handler_t (function pointer) and conn_handle
+//   * Callback is involved, use gatt_event_can_write_without_response_get_handle to get the conn_handle (no other context)
```
involved->invoked?

dpgeorge commented on extmod/btstack/modbluetooth_btstack.c:
```diff
@@ -85,6 +83,148 @@ STATIC mp_obj_bluetooth_uuid_t create_mp_uuid(uint16_t uuid16, const uint8_t *uu
+    // context_registration.context will point back to this struct.
+    btstack_context_callback_registration_t context_registration;
+
+    uint16_t value_handle;
```
Maybe shift this to after `conn_handle` so on 64-bit archs it doesn't increase size of this struct?

dpgeorge commented on extmod/btstack/modbluetooth_btstack.c:
```diff
@@ -85,6 +83,148 @@ STATIC mp_obj_bluetooth_uuid_t create_mp_uuid(uint16_t uuid16, const uint8_t *uu
+        DEBUG_EVENT_printf("btstack_notify_indicate_ready_handler: sending notification err=%d\n", err);
+        assert(err == ERROR_CODE_SUCCESS);
+        (void)err;
+    } else if (pending_op->op_type == MP_BLUETOOTH_BTSTACK_PENDING_INDICATE) {
```
Code size is decreased and slightly simplified if this is just an `else` and then in the body it does:
```c
assert(pending_op->op_type == MP_BLUETOOTH_BTSTACK_PENDING_INDICATE);
```
Then remove the `else` following it.

dpgeorge commented on extmod/btstack/modbluetooth_btstack.c:
```diff
@@ -85,6 +83,148 @@ STATIC mp_obj_bluetooth_uuid_t create_mp_uuid(uint16_t uuid16, const uint8_t *uu
+    }
+    assert(btstack_linked_list_remove(&MP_STATE_PORT(bluetooth_btstack_root_pointers)->pending_ops, (btstack_linked_item_t*)pending_op));
+    MICROPY_PY_BLUETOOTH_EXIT
+}
```
Is it possible to explicitly `m_del()` the `pending_op->buf` here, because we were responsible for allocating it and know it'll never be used again?

jimmo replied:
The buf is inline in the pending_op struct. But we can del the whole op.

I've split "remove (and optionally del)" from "find a pending op". Anywhere that attempts to unref a pending op, that isn't in interrupt context, now also frees it.

dpgeorge commented on extmod/btstack/modbluetooth_btstack.c:
```diff
@@ -85,6 +83,148 @@ STATIC mp_obj_bluetooth_uuid_t create_mp_uuid(uint16_t uuid16, const uint8_t *uu
+// Register a pending background operation -- copies the buffer, and makes it known to the GC.
+STATIC mp_btstack_pending_op_t* btstack_enqueue_pending_operation(uint16_t op_type, uint16_t conn_handle, uint16_t value_handle, const uint8_t *buf, size_t len) {
+    DEBUG_EVENT_printf("btstack_enqueue_pending_operation op_type=%d conn_handle=%d value_handle=%d len=%lu\n", op_type, conn_handle, value_handle, len);
+    mp_btstack_pending_op_t *pending_op = m_malloc(sizeof(mp_btstack_pending_op_t) + len);
```
better to use `m_new_var(mp_btstack_pending_op_t, uint8_t, len)`?

dpgeorge commented on extmod/btstack/modbluetooth_btstack.c:
```diff
@@ -85,6 +83,148 @@ STATIC mp_obj_bluetooth_uuid_t create_mp_uuid(uint16_t uuid16, const uint8_t *uu
+            DEBUG_EVENT_printf("btstack_finish_pending_operation: found value_handle=%d len=%lu\n", pending_op->value_handle, pending_op->len);
+            assert(btstack_linked_list_remove(&MP_STATE_PORT(bluetooth_btstack_root_pointers)->pending_ops, (btstack_linked_item_t*)pending_op));
+            MICROPY_PY_BLUETOOTH_EXIT
+            return pending_op;
```
as above, is it possible to explicitly free the `buf` before returning?

jimmo replied:
The caller of this function needs it (e.g. `event_type == GATT_EVENT_CAN_WRITE_WITHOUT_RESPONSE` in the packet handler). But all the callers can free it.

dpgeorge commented on extmod/btstack/modbluetooth_btstack.c:
```diff
@@ -161,12 +301,17 @@ STATIC void btstack_packet_handler(uint8_t packet_type, uint8_t *packet, uint8_t
+            mp_bluetooth_gattc_on_read_write_status(irq, conn_handle, 0xffff, status);
+            // Unref the saved buffer for the write operation on this conn_handle.
+            if (irq == MP_BLUETOOTH_IRQ_GATTC_WRITE_DONE) {
+                assert(btstack_finish_pending_operation(MP_BLUETOOTH_BTSTACK_PENDING_WRITE, conn_handle, 0xffff));
```
don't put code that must be executed in `assert()`, because it won't be executed with `NDEBUG`

dpgeorge commented on extmod/btstack/modbluetooth_btstack.c:
```diff
@@ -612,14 +769,44 @@ int mp_bluetooth_gatts_notify(uint16_t conn_handle, uint16_t value_handle) {
     DEBUG_EVENT_printf("mp_bluetooth_gatts_notify_send\n");
-    // TODO: We need to use att_server_request_to_send_notification here as the stack may not be ready to send a notification.
+
+    // Attempt to send immediately, will copy buffer.
```
"if it succeeds it will copy the buffer"

dpgeorge commented on extmod/btstack/modbluetooth_btstack.c:
```diff
@@ -612,14 +769,44 @@ int mp_bluetooth_gatts_notify(uint16_t conn_handle, uint16_t value_handle) {
+
+        att_server_request_to_send_notification(&pending_op->context_registration, conn_handle);
+        // We don't know how many bytes will be sent.
+        *value_len = 0;
```
hmm, returning 0 here means the caller always needs to test for this value to know if the notify was sent immediately or is pending... but there's really no other way to do it.  But the docs don't actually specify any return value for `gatts_notify()` so maybe it should just return `None` in all cases?

jimmo replied:
Yes, this was never reliably used anywhere. I've made `gatts_notify` always return `None` as suggested.

dpgeorge commented on extmod/btstack/modbluetooth_btstack.c:
```diff
@@ -612,14 +769,44 @@ int mp_bluetooth_gatts_notify(uint16_t conn_handle, uint16_t value_handle) {
+    }
 }
 
 int mp_bluetooth_gatts_indicate(uint16_t conn_handle, uint16_t value_handle) {
```
I just noticed: it doesn't look like this function is ever used??

jimmo replied:
I've added `gatts_indicate(conn_handle, value_handle)` to modbluetooth. (And added to the ble_characteristic.py multitest)

dpgeorge commented on extmod/btstack/modbluetooth_btstack.c:
```diff
@@ -751,19 +938,40 @@ int mp_bluetooth_gattc_read(uint16_t conn_handle, uint16_t value_handle) {
+            // Can't send right now, need to take a copy of the buffer and add it to the queue.
+            btstack_enqueue_pending_operation(MP_BLUETOOTH_BTSTACK_PENDING_WRITE_NO_RESPONSE, conn_handle, value_handle, value, *value_len);
+            // Notify when this conn_handle can write.
+            err = gatt_client_request_can_write_without_response_event(&btstack_packet_handler_generic, conn_handle);
```
if this returns an error, shouldn't `btstack_finish_pending_operation()` be called to unref the buffer?

dpgeorge commented on extmod/btstack/modbluetooth_btstack.c:
```diff
@@ -612,14 +769,44 @@ int mp_bluetooth_gatts_notify(uint16_t conn_handle, uint16_t value_handle) {
+        // Schedule callback, making a copy of the buffer.
+        mp_btstack_pending_op_t *pending_op = btstack_enqueue_pending_operation(MP_BLUETOOTH_BTSTACK_PENDING_NOTIFY, conn_handle, value_handle, value, *value_len);
+
+        att_server_request_to_send_notification(&pending_op->context_registration, conn_handle);
```
if this returns an error, shouldn't btstack_finish_pending_operation() be called to unref the buffer?

dpgeorge commented on extmod/btstack/modbluetooth_btstack.c:
```diff
@@ -612,14 +769,44 @@ int mp_bluetooth_gatts_notify(uint16_t conn_handle, uint16_t value_handle) {
+        DEBUG_EVENT_printf("mp_bluetooth_gatts_indicate: ACL buffer full, scheduling callback\n");
+        // Schedule callback, making a copy of the buffer.
+        mp_btstack_pending_op_t *pending_op = btstack_enqueue_pending_operation(MP_BLUETOOTH_BTSTACK_PENDING_INDICATE, conn_handle, value_handle, NULL, 0);
+        att_server_request_to_send_indication(&pending_op->context_registration, conn_handle);
```
if this returns an error, shouldn't btstack_finish_pending_operation() be called to unref the buffer?  In other words, can there be multiple outstanding pending ops in the queue for the same conn/value?

dpgeorge commented on extmod/btstack/modbluetooth_btstack.c:
```diff
@@ -751,19 +938,40 @@ int mp_bluetooth_gattc_read(uint16_t conn_handle, uint16_t value_handle) {
+        }
+    } else if (mode == MP_BLUETOOTH_WRITE_MODE_WITH_RESPONSE) {
+        // Pending operation copies the value buffer and keeps a GC reference until the response comes back.
+        // TODO: Is there always a response?
```
@mringwal is it guaranteed that BTstack will always issue, at some point in the future, a `GATT_EVENT_QUERY_COMPLETE` after a successful call to `gatt_client_write_value_of_characteristic` (via the callback `btstack_packet_handler_write_with_response`)?

mringwal replied:
After a successfull call to gatt_client_write_value_of_characteristic, BTstack is expected to emit a GATT_EVENT_QUERY_COMPLETE (even if the connection breaks away, it should emit it with an error status)

dpgeorge replied:
ok, thanks, it should work as expected then and we can remove this "TODO" comment

dpgeorge commented:
> Call notify/indicate at any time.

Does this mean you can queue an arbitrary number of notify/indicate data packets?

jimmo committed:
WIP: Address review comments.

jimmo commented:
Added fixes as a new commit so they can be reviewed independently. Will rebase and squash into the other commits after review.

dpgeorge commented:
Thanks for updating, it all looks very good.  Only minor thing I noticed is that the new `gatts_indicate(...)` function needs to be added to the docs.

jimmo committed:
docs: Add `gatts_indicate` to ubluetooth.rst.

Also clarify behavior of `gatts_notify` and add some TODOs about adding an
event for indication acknowledgement.

dpgeorge commented:
Thanks for adding docs.  Now rebased and merged in 5d0be97bd9d8e9511852d2917fe0c6560232e85e through b7698841b20cd20a6729ba9d9f19a93f4fd33a5c
